// web-push-notifications/lib/constants.js
'use strict';

exports.__esModule = true;
var keyTagsWasSetted = exports.keyTagsWasSetted = 'pushwooshDefaultTagsWasSetted';
var keyWasRegistered = exports.keyWasRegistered = 'pushwooshWasRegistered';
var keyApplicationCode = exports.keyApplicationCode = 'pushwooshApplicationCode';

var defaultWorkerUrl = exports.defaultWorkerUrl = '/pushwoosh-service-worker-light.js';
var defaultWorkerSecondUrl = exports.defaultWorkerSecondUrl = '/pushwoosh-service-worker-dark.js';

var keyDefaultNotificationTitle = exports.keyDefaultNotificationTitle = 'pushwooshDefaultNotificationTitle';
var keyDefaultNotificationImage = exports.keyDefaultNotificationImage = 'pushwooshDefaultNotificationImage';
var keyDefaultNotificationUrl = exports.keyDefaultNotificationUrl = 'pushwooshDefaultNotificationUrl';
var defaultNotificationTitle = exports.defaultNotificationTitle = 'Pushwoosh notification';
var defaultNotificationImage = exports.defaultNotificationImage = 'https://cp.pushwoosh.com/img/logo-medium.png';
var defaultNotificationUrl = exports.defaultNotificationUrl = '/';

var keyWorkerSDKVersion = exports.keyWorkerSDKVersion = 'pushwooshWorkerSDKVersion';

var keyLanguage = exports.keyLanguage = 'pushwooshLanguage';

var keyLastSentAppOpen = exports.keyLastSentAppOpen = 'pushwooshLastSentAppOpen';
