// web-push-notifications/lib/classes/Logger.js
'use strict';

exports.__esModule = true;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var levels = {
  error: 1,
  info: 2,
  debug: 3
};

var Logger = function () {
  function Logger(level) {
    var _this = this;

    _classCallCheck(this, Logger);

    this.setLevel(level);

    Object.keys(levels).forEach(function (k) {
      var n = levels[k];
      _this[k] = function () {
        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }

        if (n <= _this.n) {
          var _console;

          (_console = console).info.apply(_console, [k].concat(args)); // eslint-disable-line no-console
          console.trace('trace'); // eslint-disable-line no-console
        }
      };
    });
  }

  Logger.prototype.setLevel = function setLevel(level) {
    if (!levels[level]) {
      level = 'info'; // eslint-disable-line no-param-reassign
    }
    this.n = levels[level];
  };

  return Logger;
}();

exports.default = Logger;
